controllers/storageclassclaim_migration_controller.go

```diff
@@ -0,0 +1,117 @@
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+const (
+	// for migration of storageclassclaims to storageclaims
+	storageClassClaimFinalizer = "storageclassclaim.ocs.openshift.io"
+)
+
+// StorageClassClaimReconcile migrates StorageClassClaim objects to StorageClaim
+type StorageClassClaimMigrationReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+
+	log logr.Logger
+	ctx context.Context
+}
+
+func (r *StorageClassClaimMigrationReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1alpha1.StorageClassClaim{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		Complete(r)
+}
+
+//+kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclassclaims,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclassclaims/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclassclaims/finalizers,verbs=update
+
+func (r *StorageClassClaimMigrationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+
+	r.log = log.FromContext(ctx)
+	r.ctx = ctx
+	r.log.Info("Starting reconcile.")
+
+	storageClassClaim := &v1alpha1.StorageClassClaim{}
+	storageClassClaim.Name = req.Name
+
+	if err := r.get(storageClassClaim); err != nil && !kerrors.IsNotFound(err) {
+		r.log.Error(err, fmt.Sprintf("failed to get storageclassclaim %q", storageClassClaim.Name))
+		return ctrl.Result{}, err
+	}
+
+	storageClaim := &v1alpha1.StorageClaim{}
+	storageClaim.Name = storageClassClaim.Name
+
+	// blockpool (trim "pool") -> block and sharedfilesystem (trim "pool") -> sharefilesystem
+	storageClaim.Spec.Type = strings.TrimSuffix(storageClassClaim.Spec.Type, "pool")
+
+	storageClaim.Spec.EncryptionMethod = storageClassClaim.Spec.EncryptionMethod
+	storageClaim.Spec.StorageProfile = storageClassClaim.Spec.StorageProfile
+	storageClaim.Spec.StorageClient = storageClassClaim.Spec.StorageClient.DeepCopy()
+
+	r.log.Info(fmt.Sprintf("Migrating storageclassclaim %q", storageClassClaim.Name))
+	if err := r.create(storageClaim); err != nil && !kerrors.IsAlreadyExists(err) {
+		return ctrl.Result{}, fmt.Errorf("failed to create storageclaims %q: %v", storageClaim.Name, err)
+	}
+
+	for idx := range storageClassClaim.Status.SecretNames {
+		secret := &corev1.Secret{}
+		secret.Name = storageClassClaim.Status.SecretNames[idx]
+		secret.Namespace = storageClassClaim.Spec.StorageClient.Namespace
+		if err := r.delete(secret); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to delete secret %s: %v", client.ObjectKeyFromObject(secret), err)
+		}
+	}
```

Comment:
Do we really need to delete these? Isn't the storage controller reconcile them and will add an owner ref?

Reply:
I was doing the same earlier, like transfer the secret but had to do this based on your earlier review https://github.com/red-hat-storage/ocs-client-operator/pull/116#discussion_r1533902606

Reply:
Ok